### PR TITLE
JAMES-3491: JMAP PUSH should support pushState

### PR DESCRIPTION
**Pull request by LanKhuat**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala [hunk 1]
```diff
@@ -501,6 +503,47 @@ trait WebSocketContract {
...
+        })
+        .send(backend)
+        .body
+
+    Thread.sleep(100)
+
+    val jmapGuiceProbe: JmapGuiceProbe = server.getProbe(classOf[JmapGuiceProbe])
+    val emailState: State = jmapGuiceProbe.getLatestEmailState(accountId)
+    val mailboxState: State = jmapGuiceProbe.getLatestMailboxState(accountId)
+    val pushState: PushState = PushState.from(mailboxState, emailState)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala:540]: s/pushState/globalState/ ? 

If find it better describes what it is...

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala [hunk 2]
```diff
@@ -501,6 +503,47 @@ trait WebSocketContract {
...
+    val pushEnableResponse: String = s"""{"@type":"StateChange","changed":{"$ACCOUNT_ID":{"Mailbox":"${mailboxState.getValue}","Email":"${emailState.getValue}"}},"pushState":"${pushState.value}"}"""
+
+    assertThat(response.toOption.get)
+      .isEqualTo(pushEnableResponse)
+  }
+
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/WebSocketContract.scala:546]: I do belive we should add `pushState` property to all outgoing SateChanges emitted, thus we should have a code change in every test of this class...

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/change/StateChange.scala [hunk 3]
```diff
@@ -61,7 +61,8 @@ case class StateChangeEvent(eventId: EventId,
       success => success) ->
       TypeState(
         MailboxTypeName.asMap(mailboxState) ++
-          EmailTypeName.asMap(emailState))))
+          EmailTypeName.asMap(emailState))),
+      None)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/change/StateChange.scala:65]: Here it should not be 'None'... We should likely compute the global state of the server.

Maybe the 'globalState' is a compulsory property that we want to add to the SateChangeEvent?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/WebSocketTransport.scala [hunk 4]
```diff
@@ -33,14 +37,20 @@ case class WebSocketResponse(requestId: Option[RequestId], responseObject: Respo
 
 case class WebSocketError(requestId: Option[RequestId], problemDetails: ProblemDetails) extends WebSocketOutboundMessage
 
-case class StateChange(changes: Map[AccountId, TypeState]) extends WebSocketOutboundMessage {
+object PushState {
+  def from(states: State*): PushState = PushState(Hashing.sha256().hashString(states.map(_.getValue).mkString("_"), StandardCharsets.UTF_8).toString)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/core/WebSocketTransport.scala:41]: I'm not okay with this line as it is ordering dependant:

`from(emailSate, mailboxState)` will yield a different result than `from(mailboxState, emailState)`

We do not commute and this seems like an important property to me...

I suggest instead:

```
def from(emailState: State, mailboxSatate: State)
```

No ambiguity possible, that might avoid us some bugs down the line...

**Comment by chibenwa** [comment]
Needs a rebase too :-p

**Review by Arsnael — COMMENTED** [review]
Reviewed, nothing to add

**Comment by chibenwa** [comment]
Can you rebase this?

**Review by chibenwa — COMMENTED** [review]
Looks good to me

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Comment by LanKhuat** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] [ERROR] Failures: 
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] [ERROR]   StateChangeListenerTest.reactiveEventShouldOmitUnwantedTypes:73 
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] Expecting:
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3]   [StateChange(Map(AccountId(81b637d8fcd2c6da6359e6963113a1170de795e4b725b84d1e0b4cfd9ec58ce9) -> TypeState(Map(MailboxTypeName -> State(2f9f1b12-b35a-43e6-9af2-0106fb53a943)))),Some(PushState(51140cc8251eab0cffa19dbfa4f3d2b584b1447beb37b9dc4474952b7c5c06b4)))]
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] to contain exactly (and in same order):
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3]   [StateChange(Map(AccountId(81b637d8fcd2c6da6359e6963113a1170de795e4b725b84d1e0b4cfd9ec58ce9) -> TypeState(Map(MailboxTypeName -> State(2f9f1b12-b35a-43e6-9af2-0106fb53a943)))),None)]
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] but some elements were not found:
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3]   [StateChange(Map(AccountId(81b637d8fcd2c6da6359e6963113a1170de795e4b725b84d1e0b4cfd9ec58ce9) -> TypeState(Map(MailboxTypeName -> State(2f9f1b12-b35a-43e6-9af2-0106fb53a943)))),None)]
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] and others were not expected:
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3]   [StateChange(Map(AccountId(81b637d8fcd2c6da6359e6963113a1170de795e4b725b84d1e0b4cfd9ec58ce9) -> TypeState(Map(MailboxTypeName -> State(2f9f1b12-b35a-43e6-9af2-0106fb53a943)))),Some(PushState(51140cc8251eab0cffa19dbfa4f3d2b584b1447beb37b9dc4474952b7c5c06b4)))]
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] 
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] [ERROR]   StateChangeListenerTest.reactiveEventShouldSendAnOutboundMessage:54 
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] Expecting:
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3]   [StateChange(Map(AccountId(81b637d8fcd2c6da6359e6963113a1170de795e4b725b84d1e0b4cfd9ec58ce9) -> TypeState(Map(MailboxTypeName -> State(2f9f1b12-b35a-43e6-9af2-0106fb53a943), EmailTypeName -> State(2d9f1b12-b35a-43e6-9af2-0106fb53a943)))),Some(PushState(51140cc8251eab0cffa19dbfa4f3d2b584b1447beb37b9dc4474952b7c5c06b4)))]
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] to contain exactly (and in same order):
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3]   [StateChange(Map(AccountId(81b637d8fcd2c6da6359e6963113a1170de795e4b725b84d1e0b4cfd9ec58ce9) -> TypeState(Map(MailboxTypeName -> State(2f9f1b12-b35a-43e6-9af2-0106fb53a943), EmailTypeName -> State(2d9f1b12-b35a-43e6-9af2-0106fb53a943)))),None)]
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] but some elements were not found:
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3]   [StateChange(Map(AccountId(81b637d8fcd2c6da6359e6963113a1170de795e4b725b84d1e0b4cfd9ec58ce9) -> TypeState(Map(MailboxTypeName -> State(2f9f1b12-b35a-43e6-9af2-0106fb53a943), EmailTypeName -> State(2d9f1b12-b35a-43e6-9af2-0106fb53a943)))),None)]
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3] and others were not expected:
[a9e526f2ad7ed47ca830427519b73e5b6d9898a3]   [StateChange(Map(AccountId(81b637d8fcd2c6da6359e6963113a1170de795e4b725b84d1e0b4cfd9ec58ce9) -> TypeState(Map(MailboxTypeName -> State(2f9f1b12-b35a-43e6-9af2-0106fb53a943), EmailTypeName -> State(2d9f1b12-b35a-43e6-9af2-0106fb53a943)))),Some(PushState(51140cc8251eab0cffa19dbfa4f3d2b584b1447beb37b9dc4474952b7c5c06b4)))]
```
Isn't that related?

**Comment by LanKhuat** [comment]
yeah i'm having a look at it, thanks

**Comment by LanKhuat** [comment]
`
[1b9061b52a69ed827fcc580991b4d5aa140736c5] [ERROR]   CassandraMailboxMapperConcurrencyTest » NoHostAvailable All host(s) tried for ...
[1b9061b52a69ed827fcc580991b4d5aa140736c5] [ERROR]   CassandraMailboxPathV3DAOTest » NoHostAvailable All host(s) tried for query fa...
[1b9061b52a69ed827fcc580991b4d5aa140736c5] [ERROR]   CassandraMessageIdToImapUidDAOTest » NoHostAvailable All host(s) tried for que...
[1b9061b52a69ed827fcc580991b4d5aa140736c5] [ERROR]   AclV2MigrationTest » NoHostAvailable All host(s) tried for query failed (tried...
[1b9061b52a69ed827fcc580991b4d5aa140736c5] [ERROR]   SolveMailboxInconsistenciesServiceTest » NoHostAvailable All host(s) tried for...
[1b9061b52a69ed827fcc580991b4d5aa140736c5] [ERROR]   SolveMessageInconsistenciesServiceTest » NoHostAvailable All host(s) tried for...
[1b9061b52a69ed827fcc580991b4d5aa140736c5] [INFO] 
[1b9061b52a69ed827fcc580991b4d5aa140736c5] [ERROR] Tests run: 1203, Failures: 0, Errors: 43, Skipped: 27
`

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**